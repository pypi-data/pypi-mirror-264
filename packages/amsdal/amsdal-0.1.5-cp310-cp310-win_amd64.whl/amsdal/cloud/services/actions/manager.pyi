from amsdal.cloud.models.base import ListDependenciesDetails as ListDependenciesDetails, ListDeployResponse as ListDeployResponse, ListSecretsDetails as ListSecretsDetails, UpdateDeployStatusResponse as UpdateDeployStatusResponse
from amsdal.cloud.services.actions.add_dependency import AddDepenencyAction as AddDepenencyAction
from amsdal.cloud.services.actions.add_secret import AddSecretAction as AddSecretAction
from amsdal.cloud.services.actions.create_deploy import CreateDeployAction as CreateDeployAction
from amsdal.cloud.services.actions.delete_dependency import DeleteDepenencyAction as DeleteDepenencyAction
from amsdal.cloud.services.actions.delete_secret import DeleteSecretAction as DeleteSecretAction
from amsdal.cloud.services.actions.destroy_deploy import DestroyDeployAction as DestroyDeployAction
from amsdal.cloud.services.actions.expose_db import ExposeDBAction as ExposeDBAction
from amsdal.cloud.services.actions.list_dependencies import ListDependenciesAction as ListDependenciesAction
from amsdal.cloud.services.actions.list_deploys import ListDeploysAction as ListDeploysAction
from amsdal.cloud.services.actions.list_secrets import ListSecretsAction as ListSecretsAction
from amsdal.cloud.services.actions.update_deploy import UpdateDeployAction as UpdateDeployAction
from typing import Any

class CloudActionsManager:
    def create_deploy(self, deploy_type: str, lakehouse_type: str, application_uuid: str | None = None, application_name: str | None = None) -> bool: ...
    def list_deploys(self, *, list_all: bool = True) -> ListDeployResponse: ...
    def destroy_deploy(self, deployment_id: str) -> bool: ...
    def update_deploy(self, deployment_id: str) -> UpdateDeployStatusResponse: ...
    def add_secret(self, secret_name: str, secret_value: str, application_uuid: str | None = None, application_name: str | None = None) -> bool: ...
    def list_secrets(self, application_uuid: str | None = None, application_name: str | None = None, *, with_values: bool = False) -> ListSecretsDetails: ...
    def delete_secret(self, secret_name: str, application_uuid: str | None = None, application_name: str | None = None) -> bool: ...
    def add_dependency(self, dependency_name: str, application_uuid: str | None = None, application_name: str | None = None) -> bool: ...
    def delete_dependency(self, dependency_name: str, application_uuid: str | None = None, application_name: str | None = None) -> bool: ...
    def list_dependencies(self, application_uuid: str | None = None, application_name: str | None = None) -> ListDependenciesDetails: ...
    def expose_db(self, application_uuid: str | None = None, application_name: str | None = None, ip_address: str | None = None) -> dict[str, Any]: ...
