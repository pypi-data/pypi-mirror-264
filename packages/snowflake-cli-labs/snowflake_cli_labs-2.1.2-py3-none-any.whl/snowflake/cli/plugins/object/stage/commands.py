from __future__ import annotations

from pathlib import Path

import click
import typer
from snowflake.cli.api.commands.snow_typer import SnowTyper
from snowflake.cli.api.output.types import (
    CommandResult,
    ObjectResult,
    QueryResult,
    SingleQueryResult,
)
from snowflake.cli.plugins.object.stage.diff import DiffResult
from snowflake.cli.plugins.object.stage.manager import StageManager

app = SnowTyper(
    name="stage",
    help="Manages stages.",
)

StageNameArgument = typer.Argument(..., help="Name of the stage.")


@app.command("list", requires_connection=True)
def stage_list(stage_name: str = StageNameArgument, **options) -> CommandResult:
    """
    Lists the stage contents.
    """
    cursor = StageManager().list_files(stage_name=stage_name)
    return QueryResult(cursor)


def _is_stage_path(path: str):
    return path.startswith("@") or path.startswith("snow://")


@app.command("copy", requires_connection=True)
def copy(
    source_path: str = typer.Argument(
        help="Source path for copy operation. Can be either stage path or local."
    ),
    destination_path: str = typer.Argument(
        help="Target directory path for copy operation. Should be stage if source is local or local if source is stage.",
    ),
    overwrite: bool = typer.Option(
        False,
        help="Overwrites existing files in the target path.",
    ),
    parallel: int = typer.Option(
        4,
        help="Number of parallel threads to use when uploading files.",
    ),
    **options,
) -> CommandResult:
    """
    Copies all files from target path to target directory. This works for both uploading
    to and downloading files from the stage.
    """
    is_get = _is_stage_path(source_path)
    is_put = _is_stage_path(destination_path)

    if is_get and is_put:
        raise click.ClickException(
            "Both source and target path are remote. This operation is not supported."
        )
    if not is_get and not is_put:
        raise click.ClickException(
            "Both source and target path are local. This operation is not supported."
        )

    if is_get:
        target = Path(destination_path).resolve()
        cursor = StageManager().get(
            stage_name=source_path, dest_path=target, parallel=parallel
        )
    else:
        source = Path(source_path).resolve()
        local_path = str(source) + "/*" if source.is_dir() else str(source)

        cursor = StageManager().put(
            local_path=local_path,
            stage_path=destination_path,
            overwrite=overwrite,
            parallel=parallel,
        )
    return QueryResult(cursor)


@app.command("create", requires_connection=True)
def stage_create(stage_name: str = StageNameArgument, **options) -> CommandResult:
    """
    Creates a named stage if it does not already exist.
    """
    cursor = StageManager().create(stage_name=stage_name)
    return SingleQueryResult(cursor)


@app.command("remove", requires_connection=True)
def stage_remove(
    stage_name: str = StageNameArgument,
    file_name: str = typer.Argument(..., help="Name of the file to remove."),
    **options,
) -> CommandResult:
    """
    Removes a file from a stage.
    """

    cursor = StageManager().remove(stage_name=stage_name, path=file_name)
    return SingleQueryResult(cursor)


@app.command("diff", hidden=True, requires_connection=True)
def stage_diff(
    stage_name: str = typer.Argument(None, help="Fully qualified name of a stage"),
    folder_name: str = typer.Argument(None, help="Path to local folder"),
    **options,
) -> ObjectResult:
    """
    Diffs a stage with a local folder.
    """
    diff: DiffResult = stage_diff(Path(folder_name), stage_name)
    return ObjectResult(str(diff))
