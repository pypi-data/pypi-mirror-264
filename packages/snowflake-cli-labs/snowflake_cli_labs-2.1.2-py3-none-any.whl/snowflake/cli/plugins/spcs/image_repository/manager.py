from urllib.parse import urlparse

from click import ClickException
from snowflake.cli.api.constants import ObjectType
from snowflake.cli.api.project.util import (
    is_valid_unquoted_identifier,
)
from snowflake.cli.api.sql_execution import SqlExecutionMixin
from snowflake.cli.plugins.spcs.common import handle_object_already_exists
from snowflake.connector.errors import ProgrammingError


class ImageRepositoryManager(SqlExecutionMixin):
    def get_database(self):
        return self._conn.database

    def get_schema(self):
        return self._conn.schema

    def get_role(self):
        return self._conn.role

    def get_repository_url(self, repo_name: str, with_scheme: bool = True):
        if not is_valid_unquoted_identifier(repo_name):
            raise ValueError(
                f"repo_name '{repo_name}' is not a valid unquoted Snowflake identifier"
            )
        # we explicitly do not allow this function to be used without connection database and schema set
        repo_row = self.show_specific_object(
            "image repositories", repo_name, check_schema=True
        )
        if repo_row is None:
            raise ClickException(
                f"Image repository '{repo_name}' does not exist in database '{self.get_database()}' and schema '{self.get_schema()}' or not authorized."
            )
        if with_scheme:
            return f"https://{repo_row['repository_url']}"
        else:
            return repo_row["repository_url"]

    def get_repository_api_url(self, repo_url):
        """
        Converts a repo URL to a registry OCI API URL.
        https://reg.com/db/schema/repo becomes https://reg.com/v2/db/schema/repo
        """
        parsed_url = urlparse(repo_url)

        scheme = parsed_url.scheme
        host = parsed_url.netloc
        path = parsed_url.path

        return f"{scheme}://{host}/v2{path}"

    def create(self, name: str):
        try:
            return self._execute_schema_query(f"create image repository {name}")
        except ProgrammingError as e:
            handle_object_already_exists(e, ObjectType.IMAGE_REPOSITORY, name)
