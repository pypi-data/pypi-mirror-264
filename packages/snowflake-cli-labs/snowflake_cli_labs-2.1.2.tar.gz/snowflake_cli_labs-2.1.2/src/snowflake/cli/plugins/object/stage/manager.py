from __future__ import annotations

import logging
import re
from contextlib import nullcontext
from pathlib import Path
from typing import Optional, Union

from snowflake.cli.api.project.util import to_string_literal
from snowflake.cli.api.sql_execution import SqlExecutionMixin
from snowflake.cli.api.utils.path_utils import path_resolver
from snowflake.connector.cursor import SnowflakeCursor

log = logging.getLogger(__name__)


UNQUOTED_FILE_URI_REGEX = r"[\w/*?\-.=&{}$#[\]\"\\!@%^+:]+"


class StageManager(SqlExecutionMixin):
    @staticmethod
    def get_standard_stage_name(name: str) -> str:
        # Handle embedded stages
        if name.startswith("snow://") or name.startswith("@"):
            return name

        return f"@{name}"

    @staticmethod
    def get_stage_name_from_path(path: str):
        """
        Returns stage name from potential path on stage. For example
        db.schema.stage/foo/bar  -> db.schema.stage
        """
        return Path(path).parts[0]

    @staticmethod
    def quote_stage_name(name: str) -> str:
        if name.startswith("'") and name.endswith("'"):
            return name  # already quoted

        standard_name = StageManager.get_standard_stage_name(name)
        if standard_name.startswith("@") and not re.fullmatch(
            r"@([\w./$])+", standard_name
        ):
            return to_string_literal(standard_name)

        return standard_name

    def _to_uri(self, local_path: str):
        uri = f"file://{local_path}"
        if re.fullmatch(UNQUOTED_FILE_URI_REGEX, uri):
            return uri
        return to_string_literal(uri)

    def list_files(self, stage_name: str) -> SnowflakeCursor:
        stage_name = self.get_standard_stage_name(stage_name)
        return self._execute_query(f"ls {self.quote_stage_name(stage_name)}")

    def get(
        self, stage_name: str, dest_path: Path, parallel: int = 4
    ) -> SnowflakeCursor:
        stage_name = self.get_standard_stage_name(stage_name)
        dest_directory = f"{dest_path}/"
        return self._execute_query(
            f"get {self.quote_stage_name(stage_name)} {self._to_uri(dest_directory)} parallel={parallel}"
        )

    def put(
        self,
        local_path: Union[str, Path],
        stage_path: str,
        parallel: int = 4,
        overwrite: bool = False,
        role: Optional[str] = None,
    ) -> SnowflakeCursor:
        """
        This method will take a file path from the user's system and put it into a Snowflake stage,
        which includes its fully qualified name as well as the path within the stage.
        If provided with a role, then temporarily use this role to perform the operation above,
        and switch back to the original role for the next commands to run.
        """
        with self.use_role(role) if role else nullcontext():
            stage_path = self.get_standard_stage_name(stage_path)
            local_resolved_path = path_resolver(str(local_path))
            log.info("Uploading %s to @%s", local_resolved_path, stage_path)
            cursor = self._execute_query(
                f"put {self._to_uri(local_resolved_path)} {self.quote_stage_name(stage_path)} "
                f"auto_compress=false parallel={parallel} overwrite={overwrite}"
            )
        return cursor

    def remove(
        self, stage_name: str, path: str, role: Optional[str] = None
    ) -> SnowflakeCursor:
        """
        This method will take a file path that exists on a Snowflake stage,
        and remove it from the stage.
        If provided with a role, then temporarily use this role to perform the operation above,
        and switch back to the original role for the next commands to run.
        """
        with self.use_role(role) if role else nullcontext():
            stage_name = self.get_standard_stage_name(stage_name)
            path = path if path.startswith("/") else "/" + path
            quoted_stage_name = self.quote_stage_name(f"{stage_name}{path}")
            return self._execute_query(f"remove {quoted_stage_name}")

    def create(self, stage_name: str, comment: Optional[str] = None) -> SnowflakeCursor:
        query = f"create stage if not exists {stage_name}"
        if comment:
            query += f" comment='{comment}'"
        return self._execute_query(query)
