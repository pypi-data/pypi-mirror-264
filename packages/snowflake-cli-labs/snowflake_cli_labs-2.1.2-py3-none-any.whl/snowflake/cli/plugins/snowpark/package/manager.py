from __future__ import annotations

import logging
import os.path
from pathlib import Path

from requirements.requirement import Requirement
from snowflake.cli.api.constants import PACKAGES_DIR
from snowflake.cli.api.secure_path import SecurePath
from snowflake.cli.plugins.object.stage.manager import StageManager
from snowflake.cli.plugins.snowpark import package_utils
from snowflake.cli.plugins.snowpark.models import SplitRequirements
from snowflake.cli.plugins.snowpark.package.utils import (
    CreatedSuccessfully,
    InAnaconda,
    LookupResult,
    NothingFound,
    NotInAnaconda,
    RequiresPackages,
    prepare_app_zip,
)
from snowflake.cli.plugins.snowpark.zipper import zip_dir

log = logging.getLogger(__name__)


def lookup(name: str, install_packages: bool) -> LookupResult:

    package_response = package_utils.parse_anaconda_packages([Requirement.parse(name)])

    if package_response.snowflake and not package_response.other:
        return InAnaconda(package_response, name)
    elif install_packages:
        status, result = package_utils.install_packages(
            perform_anaconda_check=True, package_name=name, file_name=None
        )

        if status:
            if result.snowflake:
                return RequiresPackages(result, name)
            else:
                return NotInAnaconda(result, name)

    return NothingFound(SplitRequirements([], []), name)


def upload(file: Path, stage: str, overwrite: bool):
    log.info("Uploading %s to Snowflake @%s/%s...", file, stage, file)
    with SecurePath.temporary_directory() as temp_dir:
        temp_app_zip_path = prepare_app_zip(SecurePath(file), temp_dir)
        sm = StageManager()

        sm.create(sm.get_stage_name_from_path(stage))
        put_response = sm.put(
            temp_app_zip_path.path, stage, overwrite=overwrite
        ).fetchone()

    message = f"Package {file} {put_response[6]} to Snowflake @{stage}/{file}."

    if put_response[6] == "SKIPPED":
        message = "Package already exists on stage. Consider using --overwrite to overwrite the file."

    return message


def create(zip_name: str):
    file_name = zip_name if zip_name.endswith(".zip") else f"{zip_name}.zip"
    zip_dir(dest_zip=Path(file_name), source=Path.cwd() / ".packages")

    if os.path.exists(file_name):
        return CreatedSuccessfully(zip_name, Path(file_name))


def cleanup_after_install():
    if PACKAGES_DIR.exists():
        SecurePath(PACKAGES_DIR).rmdir(recursive=True)
