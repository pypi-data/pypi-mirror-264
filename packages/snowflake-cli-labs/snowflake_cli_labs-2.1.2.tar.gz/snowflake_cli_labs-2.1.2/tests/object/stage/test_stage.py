from pathlib import Path
from tempfile import TemporaryDirectory
from unittest import mock

import pytest
from snowflake.cli.plugins.object.stage.manager import StageManager
from snowflake.connector.cursor import DictCursor

STAGE_MANAGER = "snowflake.cli.plugins.object.stage.manager.StageManager"


@mock.patch(f"{STAGE_MANAGER}._execute_query")
def test_stage_list(mock_execute, runner, mock_cursor):
    mock_execute.return_value = mock_cursor(["row"], [])
    result = runner.invoke(["object", "stage", "list", "-c", "empty", "stageName"])
    assert result.exit_code == 0, result.output
    mock_execute.assert_called_once_with("ls @stageName")


@mock.patch(f"{STAGE_MANAGER}._execute_query")
def test_stage_list_quoted(mock_execute, runner, mock_cursor):
    mock_execute.return_value = mock_cursor(["row"], [])
    result = runner.invoke(["object", "stage", "list", "-c", "empty", '"stage name"'])
    assert result.exit_code == 0, result.output
    mock_execute.assert_called_once_with("ls '@\"stage name\"'")


@mock.patch(f"{STAGE_MANAGER}._execute_query")
def test_stage_copy_remote_to_local(mock_execute, runner, mock_cursor):
    mock_execute.return_value = mock_cursor(["row"], [])
    with TemporaryDirectory() as tmp_dir:
        result = runner.invoke(
            ["object", "stage", "copy", "-c", "empty", "@stageName", str(tmp_dir)]
        )
    assert result.exit_code == 0, result.output
    mock_execute.assert_called_once_with(
        f"get @stageName file://{Path(tmp_dir).resolve()}/ parallel=4"
    )


@mock.patch(f"{STAGE_MANAGER}._execute_query")
def test_stage_copy_remote_to_local_quoted_stage(mock_execute, runner, mock_cursor):
    mock_execute.return_value = mock_cursor(["row"], [])
    with TemporaryDirectory() as tmp_dir:
        result = runner.invoke(
            ["object", "stage", "copy", "-c", "empty", '@"stage name"', str(tmp_dir)]
        )
    assert result.exit_code == 0, result.output
    mock_execute.assert_called_once_with(
        f"get '@\"stage name\"' file://{Path(tmp_dir).resolve()}/ parallel=4"
    )


@pytest.mark.parametrize(
    "raw_path,expected_uri",
    [
        ("{}/dest", "file://{}/dest/"),
        ("{}/dest.dir", "file://{}/dest.dir/"),
        ("{}/dest0", "file://{}/dest0/"),
        ("{}/dest dir", "'file://{}/dest dir/'"),
        ("{}/dest#dir", "file://{}/dest#dir/"),
        ("{}/dest*dir", "file://{}/dest*dir/"),
        ("{}/dest_?dir", "file://{}/dest_?dir/"),
        ("{}/dest%dir", "file://{}/dest%dir/"),
        ('{}/dest"dir', 'file://{}/dest"dir/'),
        ("{}/dest'dir", r"'file://{}/dest\'dir/'"),
        ("{}/dest\tdir", r"'file://{}/dest\tdir/'"),
    ],
)
@mock.patch(f"{STAGE_MANAGER}._execute_query")
def test_stage_copy_remote_to_local_quoted_uri(
    mock_execute, runner, mock_cursor, raw_path, expected_uri
):
    mock_execute.return_value = mock_cursor(["row"], [])
    with TemporaryDirectory() as tmp_dir:
        tmp_dir = Path(tmp_dir).resolve()
        local_path = raw_path.replace("{}", str(tmp_dir))
        file_uri = expected_uri.replace("{}", str(tmp_dir))
        result = runner.invoke(
            ["object", "stage", "copy", "-c", "empty", "@stageName", local_path]
        )
    assert result.exit_code == 0, result.output
    mock_execute.assert_called_once_with(f"get @stageName {file_uri} parallel=4")


@mock.patch(f"{STAGE_MANAGER}._execute_query")
def test_stage_copy_local_to_remote(mock_execute, runner, mock_cursor):
    mock_execute.return_value = mock_cursor(["row"], [])
    with TemporaryDirectory() as tmp_dir:
        result = runner.invoke(
            [
                "object",
                "stage",
                "copy",
                "-c",
                "empty",
                "--overwrite",
                "--parallel",
                42,
                str(tmp_dir),
                "@stageName",
            ]
        )
    assert result.exit_code == 0, result.output
    mock_execute.assert_called_once_with(
        f"put file://{Path(tmp_dir).resolve()}/* @stageName auto_compress=false parallel=42 overwrite=True"
    )


@mock.patch(f"{STAGE_MANAGER}._execute_query")
def test_stage_copy_local_to_remote_quoted_stage(mock_execute, runner, mock_cursor):
    mock_execute.return_value = mock_cursor(["row"], [])
    with TemporaryDirectory() as tmp_dir:
        result = runner.invoke(
            [
                "object",
                "stage",
                "copy",
                "-c",
                "empty",
                "--overwrite",
                "--parallel",
                42,
                str(tmp_dir),
                '@"stage name"',
            ]
        )
    assert result.exit_code == 0, result.output
    mock_execute.assert_called_once_with(
        f"put file://{Path(tmp_dir).resolve()}/* '@\"stage name\"' auto_compress=false parallel=42 overwrite=True"
    )


@pytest.mark.parametrize(
    "raw_path,expected_uri",
    [
        ("{}/readme.md", "file://{}/readme.md"),
        ("{}/readme0.md", "file://{}/readme0.md"),
        ("{}/read me.md", "'file://{}/read me.md'"),
        ("{}/read#me.md", "file://{}/read#me.md"),
        ("{}/read*.md", "file://{}/read*.md"),
        ("{}/read_?me.md", "file://{}/read_?me.md"),
        ("{}/read%me.md", "file://{}/read%me.md"),
        ('{}/read"me.md', 'file://{}/read"me.md'),
        ("{}/read'me.md", r"'file://{}/read\'me.md'"),
        ("{}/read\tme.md", r"'file://{}/read\tme.md'"),
    ],
)
@mock.patch(f"{STAGE_MANAGER}._execute_query")
def test_stage_copy_local_to_remote_quoted_path(
    mock_execute, runner, mock_cursor, raw_path, expected_uri
):
    mock_execute.return_value = mock_cursor(["row"], [])
    with TemporaryDirectory() as tmp_dir:
        tmp_dir = Path(tmp_dir).resolve()
        local_path = raw_path.replace("{}", str(tmp_dir))
        file_uri = expected_uri.replace("{}", str(tmp_dir))

        result = runner.invoke(
            [
                "object",
                "stage",
                "copy",
                "-c",
                "empty",
                "--overwrite",
                "--parallel",
                42,
                local_path,
                "@stageName",
            ]
        )
    assert result.exit_code == 0, result.output
    mock_execute.assert_called_once_with(
        f"put {file_uri} @stageName auto_compress=false parallel=42 overwrite=True"
    )


@mock.patch(f"{STAGE_MANAGER}._execute_query")
def test_stage_copy_local_to_remote_star(mock_execute, runner, mock_cursor):
    mock_execute.return_value = mock_cursor(["row"], [])
    with TemporaryDirectory() as tmp_dir:
        result = runner.invoke(
            [
                "object",
                "stage",
                "copy",
                "-c",
                "empty",
                "--overwrite",
                "--parallel",
                42,
                str(tmp_dir) + "/*.py",
                "@stageName",
            ]
        )
    assert result.exit_code == 0, result.output
    mock_execute.assert_called_once_with(
        f"put file://{Path(tmp_dir).resolve()}/*.py @stageName auto_compress=false parallel=42 overwrite=True"
    )


@pytest.mark.parametrize(
    "source, dest",
    [
        ("@snow/stage", "@stage/snow"),
        ("snow://stage", "snow://stage/snow"),
        ("local/path", "other/local/path"),
    ],
)
def test_copy_throws_error_for_same_platform_operation(runner, source, dest, snapshot):
    result = runner.invoke(["object", "stage", "copy", source, dest])
    assert result.exit_code == 1
    assert result.output == snapshot


@mock.patch(f"{STAGE_MANAGER}._execute_query")
def test_stage_create(mock_execute, runner, mock_cursor):
    mock_execute.return_value = mock_cursor(["row"], [])
    result = runner.invoke(["object", "stage", "create", "-c", "empty", "stageName"])
    assert result.exit_code == 0, result.output
    mock_execute.assert_called_once_with("create stage if not exists stageName")


@mock.patch(f"{STAGE_MANAGER}._execute_query")
def test_stage_create_quoted(mock_execute, runner, mock_cursor):
    mock_execute.return_value = mock_cursor(["row"], [])
    result = runner.invoke(["object", "stage", "create", "-c", "empty", '"stage name"'])
    assert result.exit_code == 0, result.output
    mock_execute.assert_called_once_with('create stage if not exists "stage name"')


@mock.patch("snowflake.cli.plugins.object.commands.ObjectManager._execute_query")
def test_stage_drop(mock_execute, runner, mock_cursor):
    mock_execute.return_value = mock_cursor(["row"], [])
    result = runner.invoke(["object", "drop", "stage", "stageName", "-c", "empty"])
    assert result.exit_code == 0, result.output
    mock_execute.assert_called_once_with("drop stage stageName")


@mock.patch("snowflake.cli.plugins.object.commands.ObjectManager._execute_query")
def test_stage_drop_quoted(mock_execute, runner, mock_cursor):
    mock_execute.return_value = mock_cursor(["row"], [])
    result = runner.invoke(["object", "drop", "stage", '"stage name"', "-c", "empty"])
    assert result.exit_code == 0, result.output
    mock_execute.assert_called_once_with('drop stage "stage name"')


@mock.patch(f"{STAGE_MANAGER}._execute_query")
def test_stage_remove(mock_execute, runner, mock_cursor):
    mock_execute.return_value = mock_cursor(["row"], [])
    result = runner.invoke(
        ["object", "stage", "remove", "-c", "empty", "stageName", "my/file/foo.csv"]
    )
    assert result.exit_code == 0, result.output
    mock_execute.assert_called_once_with("remove @stageName/my/file/foo.csv")


@mock.patch(f"{STAGE_MANAGER}._execute_query")
def test_stage_remove_quoted(mock_execute, runner, mock_cursor):
    mock_execute.return_value = mock_cursor(["row"], [])
    result = runner.invoke(
        ["object", "stage", "remove", "-c", "empty", '"stage name"', "my/file/foo.csv"]
    )
    assert result.exit_code == 0, result.output
    mock_execute.assert_called_once_with("remove '@\"stage name\"/my/file/foo.csv'")


@mock.patch(f"{STAGE_MANAGER}._execute_query")
def test_stage_print_result_for_put_directory(
    mock_execute, mock_cursor, runner, snapshot
):
    mock_execute.return_value = mock_cursor(
        rows=[
            ["file1.txt", "file1.txt", 10, 8, "NONE", "NONE", "UPLOADED", ""],
            ["file2.txt", "file2.txt", 10, 8, "NONE", "NONE", "UPLOADED", ""],
            ["file3.txt", "file3.txt", 10, 8, "NONE", "NONE", "UPLOADED", ""],
        ],
        columns=[
            "source",
            "target",
            "source_size",
            "target_size",
            "source_compression",
            "target_compression",
            "status",
            "message",
        ],
    )

    with TemporaryDirectory() as tmp_dir:
        tmp_dir_path = Path(tmp_dir)
        (tmp_dir_path / "file1.txt").touch()
        (tmp_dir_path / "file2.txt").touch()
        (tmp_dir_path / "file3.txt").touch()
        result = runner.invoke(["object", "stage", "copy", tmp_dir, "@stageName"])

    assert result.exit_code == 0, result.output
    assert result.output == snapshot


@mock.patch(f"{STAGE_MANAGER}._execute_query")
def test_stage_print_result_for_get_all_files_from_stage(
    mock_execute, mock_cursor, runner, snapshot
):
    mock_execute.return_value = mock_cursor(
        rows=[
            ["file1.txt", 10, "DOWNLOADED", ""],
            ["file2.txt", 10, "DOWNLOADED", ""],
            ["file3.txt", 10, "DOWNLOADED", ""],
        ],
        columns=[
            "file",
            "size",
            "status",
            "message",
        ],
    )

    with TemporaryDirectory() as tmp_dir:
        result = runner.invoke(["object", "stage", "copy", "@stageName", tmp_dir])

    assert result.exit_code == 0, result.output
    assert result.output == snapshot


@mock.patch(f"{STAGE_MANAGER}._execute_query")
def test_stage_internal_remove(mock_execute, mock_cursor):
    mock_execute.return_value = mock_cursor([{"CURRENT_ROLE()": "old_role"}], [])
    sm = StageManager()
    sm.remove("stageName", "my/file/foo.csv", "new_role")
    expected = [
        mock.call("select current_role()", cursor_class=DictCursor),
        mock.call("use role new_role"),
        mock.call("remove @stageName/my/file/foo.csv"),
        mock.call("use role old_role"),
    ]
    assert mock_execute.mock_calls == expected


@mock.patch(f"{STAGE_MANAGER}._execute_query")
def test_stage_internal_remove_quoted(mock_execute, mock_cursor):
    mock_execute.return_value = mock_cursor([{"CURRENT_ROLE()": "old_role"}], [])
    sm = StageManager()
    sm.remove('"stage name"', "my/file/foo.csv", "new_role")
    expected = [
        mock.call("select current_role()", cursor_class=DictCursor),
        mock.call("use role new_role"),
        mock.call("remove '@\"stage name\"/my/file/foo.csv'"),
        mock.call("use role old_role"),
    ]
    assert mock_execute.mock_calls == expected


@mock.patch(f"{STAGE_MANAGER}._execute_query")
def test_stage_internal_remove_no_role_change(mock_execute, mock_cursor):
    mock_execute.return_value = mock_cursor([{"CURRENT_ROLE()": "old_role"}], [])
    sm = StageManager()
    sm.remove("stageName", "my/file/foo.csv", "old_role")
    expected = [
        mock.call("select current_role()", cursor_class=DictCursor),
        mock.call("remove @stageName/my/file/foo.csv"),
    ]
    assert mock_execute.mock_calls == expected


@mock.patch(f"{STAGE_MANAGER}._execute_query")
def test_stage_internal_put(mock_execute, mock_cursor):
    mock_execute.return_value = mock_cursor([{"CURRENT_ROLE()": "old_role"}], [])
    with TemporaryDirectory() as tmp_dir:
        sm = StageManager()
        sm.put(Path(tmp_dir).resolve(), "stageName", role="new_role")
        expected = [
            mock.call("select current_role()", cursor_class=DictCursor),
            mock.call("use role new_role"),
            mock.call(
                f"put file://{Path(tmp_dir).resolve()} @stageName auto_compress=false parallel=4 overwrite=False"
            ),
            mock.call("use role old_role"),
        ]
        assert mock_execute.mock_calls == expected


@mock.patch(f"{STAGE_MANAGER}._execute_query")
def test_stage_internal_put_quoted_stage(mock_execute, mock_cursor):
    mock_execute.return_value = mock_cursor([{"CURRENT_ROLE()": "old_role"}], [])
    with TemporaryDirectory() as tmp_dir:
        sm = StageManager()
        sm.put(Path(tmp_dir).resolve(), '"stage name"', role="new_role")
        expected = [
            mock.call("select current_role()", cursor_class=DictCursor),
            mock.call("use role new_role"),
            mock.call(
                f"put file://{Path(tmp_dir).resolve()} '@\"stage name\"' auto_compress=false parallel=4 overwrite=False"
            ),
            mock.call("use role old_role"),
        ]
        assert mock_execute.mock_calls == expected


@pytest.mark.parametrize(
    "raw_path,expected_uri",
    [
        ("{}/readme.md", "file://{}/readme.md"),
        ("{}/readme0.md", "file://{}/readme0.md"),
        ("{}/read me.md", "'file://{}/read me.md'"),
        ("{}/read#me.md", "file://{}/read#me.md"),
        ("{}/read*.md", "file://{}/read*.md"),
        ("{}/read_?me.md", "file://{}/read_?me.md"),
        ("{}/read%me.md", "file://{}/read%me.md"),
        ('{}/read"me.md', 'file://{}/read"me.md'),
        ("{}/read'me.md", r"'file://{}/read\'me.md'"),
        ("{}/read\tme.md", r"'file://{}/read\tme.md'"),
    ],
)
@mock.patch(f"{STAGE_MANAGER}._execute_query")
def test_stage_internal_put_quoted_path(
    mock_execute, mock_cursor, raw_path, expected_uri
):
    mock_execute.return_value = mock_cursor([{"CURRENT_ROLE()": "old_role"}], [])
    with TemporaryDirectory() as tmp_dir:
        sm = StageManager()
        tmp_dir = Path(tmp_dir).resolve()
        src_path = raw_path.replace("{}", str(tmp_dir))
        src_uri = expected_uri.replace("{}", str(tmp_dir))
        sm.put(src_path, "stageName", role="new_role")
        expected = [
            mock.call("select current_role()", cursor_class=DictCursor),
            mock.call("use role new_role"),
            mock.call(
                f"put {src_uri} @stageName auto_compress=false parallel=4 overwrite=False"
            ),
            mock.call("use role old_role"),
        ]
        assert mock_execute.mock_calls == expected
