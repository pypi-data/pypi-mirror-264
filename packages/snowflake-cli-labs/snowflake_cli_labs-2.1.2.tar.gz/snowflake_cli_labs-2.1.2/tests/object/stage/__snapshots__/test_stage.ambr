# serializer version: 1
# name: test_copy_throws_error_for_same_platform_operation[@snow/stage-@stage/snow]
  '''
  ╭─ Error ──────────────────────────────────────────────────────────────────────╮
  │ Both source and target path are remote. This operation is not supported.     │
  ╰──────────────────────────────────────────────────────────────────────────────╯
  
  '''
# ---
# name: test_copy_throws_error_for_same_platform_operation[local/path-other/local/path]
  '''
  ╭─ Error ──────────────────────────────────────────────────────────────────────╮
  │ Both source and target path are local. This operation is not supported.      │
  ╰──────────────────────────────────────────────────────────────────────────────╯
  
  '''
# ---
# name: test_copy_throws_error_for_same_platform_operation[snow://stage-snow://stage/snow]
  '''
  ╭─ Error ──────────────────────────────────────────────────────────────────────╮
  │ Both source and target path are remote. This operation is not supported.     │
  ╰──────────────────────────────────────────────────────────────────────────────╯
  
  '''
# ---
# name: test_stage_print_result_for_get_all_files_from_stage
  '''
  SELECT A MOCK QUERY
  +-----------------------------------------+
  | file      | size | status     | message |
  |-----------+------+------------+---------|
  | file1.txt | 10   | DOWNLOADED |         |
  | file2.txt | 10   | DOWNLOADED |         |
  | file3.txt | 10   | DOWNLOADED |         |
  +-----------------------------------------+
  
  '''
# ---
# name: test_stage_print_result_for_put_directory
  '''
  SELECT A MOCK QUERY
  +------------------------------------------------------------------------------+
  |         |         |         |         | source_ | target_ |         |        |
  |         |         | source_ | target_ | compres | compres |         | messag |
  | source  | target  | size    | size    | sion    | sion    | status  | e      |
  |---------+---------+---------+---------+---------+---------+---------+--------|
  | file1.t | file1.t | 10      | 8       | NONE    | NONE    | UPLOADE |        |
  | xt      | xt      |         |         |         |         | D       |        |
  | file2.t | file2.t | 10      | 8       | NONE    | NONE    | UPLOADE |        |
  | xt      | xt      |         |         |         |         | D       |        |
  | file3.t | file3.t | 10      | 8       | NONE    | NONE    | UPLOADE |        |
  | xt      | xt      |         |         |         |         | D       |        |
  +------------------------------------------------------------------------------+
  
  '''
# ---
